File: core/amber/src/main/python/core/architecture/packaging/tuple_to_batch_converter.py
from collections import OrderedDict
from itertools import chain

from typing import Iterable, Iterator

from core import Tuple
from core.architecture.sendsemantics.data_sending_policy_exec import DataSendingPolicyExec
from core.architecture.sendsemantics.one_to_one_policy_exec import OneToOnePolicyExec
from core.models.payload import DataFrame, DataPayload
from core.util.proto.proto_helper import get_one_of
from edu.uci.ics.amber.engine.architecture.sendsemantics import DataSendingPolicy, OneToOnePolicy
from edu.uci.ics.amber.engine.common import ActorVirtualIdentity, LinkIdentity


class TupleToBatchConverter:

    def __init__(self, ):
        self._policy_execs: OrderedDict[LinkIdentity, DataSendingPolicy] = OrderedDict()
        self._policy_exec_map: dict[type(DataSendingPolicy), type(DataSendingPolicyExec)] = {
            OneToOnePolicy: OneToOnePolicyExec
        }

    def add_policy(self, policy: DataSendingPolicy) -> None:
        """
        Add down stream operator and its transfer policy
        :param policy:
        :return:
        """
        the_policy = get_one_of(policy)
        policy_exec: type = self._policy_exec_map[type(the_policy)]
        policy_exec_instance: DataSendingPolicyExec = policy_exec(the_policy)
        self._policy_execs.update({the_policy.policy_tag: policy_exec_instance})

    def tuple_to_batch(self, tuple_: Tuple) -> Iterator[tuple[ActorVirtualIdentity, DataFrame]]:
        return chain(*(policy_exec.add_tuple_to_batch(tuple_) for policy_exec in self._policy_execs.values()))

    def emit_end_of_upstream(self) -> Iterable[tuple[ActorVirtualIdentity, DataPayload]]:
        return chain(*(policy_exec.no_more() for policy_exec in self._policy_execs.values()))

File: core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/messaginglayer/TupleToBatchConverter.scala
package edu.uci.ics.amber.engine.architecture.messaginglayer

import edu.uci.ics.amber.engine.architecture.sendsemantics.datatransferpolicy.DataSendingPolicy
import edu.uci.ics.amber.engine.common.ambermessage.DataPayload
import edu.uci.ics.amber.engine.common.tuple.ITuple
import edu.uci.ics.amber.engine.common.virtualidentity.ActorVirtualIdentity

import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer
import scala.util.control.Breaks

/** This class is a container of all the transfer policies.
  * @param selfID
  * @param dataOutputPort
  * @param controlOutputPort
  */
class TupleToBatchConverter(
    selfID: ActorVirtualIdentity,
    dataOutputPort: DataOutputPort
) {
  private var policies = new Array[DataSendingPolicy](0)

  /** Add down stream operator and its transfer policy
    * @param policy
    * @param linkTag
    * @param receivers
    */
  def addPolicy(
      policy: DataSendingPolicy
  ): Unit = {
    var i = 0
    Breaks.breakable {
      while (i < policies.length) {
        if (policies(i).policyTag == policy.policyTag) {
          policies(i) = policy
          Breaks.break()
        }
        i += 1
      }
      policies = policies :+ policy
    }
  }

  def getWorkloadHistory(): mutable.HashMap[ActorVirtualIdentity, ArrayBuffer[Long]] = {
    var ret: mutable.HashMap[ActorVirtualIdentity, ArrayBuffer[Long]] = null
    policies.foreach(policy => {
      ret = policy.getWorkloadHistory(selfID)
    })
    if (ret == null) {
      ret = new mutable.HashMap[ActorVirtualIdentity, ArrayBuffer[Long]]
    }
    ret
  }

  def recordHistory(): Unit = {
    policies.foreach(policy => {
      policy.recordHistory = true
    })
  }

  def changeFlow(
      skewedReceiverId: ActorVirtualIdentity,
      freeReceiverId: ActorVirtualIdentity,
      tuplesToRedirectNumerator: Long,
      tuplesToRedirectDenominator: Long
  ): Map[ActorVirtualIdentity, Long] = {
    var receiverToSentCount: Map[ActorVirtualIdentity, Long] = null
    policies.foreach(policy => {
      receiverToSentCount = policy.addReceiverToBucket(
        skewedReceiverId,
        freeReceiverId,
        tuplesToRedirectNumerator,
        tuplesToRedirectDenominator
      )
    })
    if (receiverToSentCount == null) {
      receiverToSentCount = new mutable.HashMap[ActorVirtualIdentity, Long]().toMap
    }
    receiverToSentCount
  }

  def rollbackFlow(
      skewedReceiverId: ActorVirtualIdentity,
      freeReceiverId: ActorVirtualIdentity
  ): Map[ActorVirtualIdentity, Long] = {
    var receiverToSentCount: Map[ActorVirtualIdentity, Long] = null
    policies.foreach(policy => {
      receiverToSentCount = policy.removeReceiverFromBucket(skewedReceiverId, freeReceiverId)
    })
    if (receiverToSentCount == null) {
      receiverToSentCount = new mutable.HashMap[ActorVirtualIdentity, Long]().toMap
    }
    receiverToSentCount
  }

  /** Push one tuple to the downstream, will be batched by each transfer policy.
    * Should ONLY be called by DataProcessor.
    * @param tuple
    */
  def passTupleToDownstream(tuple: ITuple): Unit = {
    var i = 0
    while (i < policies.length) {
      val receiverAndBatch: Option[(ActorVirtualIdentity, DataPayload)] =
        policies(i).addTupleToBatch(tuple)
      receiverAndBatch match {
        case Some((id, batch)) =>
          // send it to messaging layer to be sent downstream
          dataOutputPort.sendTo(id, batch)
        case None =>
        // Do nothing
      }
      i += 1
    }
  }

  /* Old API: for compatibility */
  def resetPolicies(): Unit = {
    policies.foreach(_.reset())
  }

  /* Send the last batch and EOU marker to all down streams */
  def emitEndOfUpstream(): Unit = {
    var i = 0
    while (i < policies.length) {
      val receiversAndBatches: Array[(ActorVirtualIdentity, DataPayload)] = policies(i).noMore()
      receiversAndBatches.foreach {
        case (id, batch) => dataOutputPort.sendTo(id, batch)
      }
      i += 1
    }
  }

}
